### PR TITLE
core: batch db operations for group leave and delete

### PR DESCRIPTION
**Pull request by spaced4ndy**
Changes in this PR:
- batches db operations for connection updates when deleting member connections
- batches db operations for unused contact deletion
- reworks bulk file cancellation and deletion
  - cancellation and deletion are split into different functions, some call sites only do cancellation (e.g. leave group), some others only do deletion (e.g. clear notes)
  - leave group api now does file cancellation
  - file cancellation db operations are batched in steps (getFileTransfer, update files to cancelled, set xftp files "deleted" flags)
  - other call sites in addition to group leave and delete also use reworked approach: clear chat, user deletion, chat item expiration, some call sites deleting only a single file (e.g. chat item deletion)
  - some other call sites deleting only a single file continue to use specialized functions such as cancelRcvFileTransfer: SMP file connections, cancel file api (the latter makes more detailed checks and some additional actions e.g. to reset file for re-download); can be further simplified if SMP and inline files protocols are deprecated
  - in batch cancellation FileChunkCancel is no longer sent for SMP and inline files for simplicity (as they are no longer in use and may be deprecated soon)

### CODE REVIEW

File: src/Simplex/Chat.hs [hunk 1]
```diff
@@ -2628,20 +2634,101 @@ setAllExpireCIFlags b = do
...
+            Just xrf -> (sndXFTPs, CancelRcvXFTPFile ft xrf : rcvXFTPs, sndSMPs, rcvSMPs)
+            Nothing -> (sndXFTPs, rcvXFTPs, sndSMPs, CancelRcvSMPFile ft : rcvSMPs)
+    cancelSndXFTPs :: [CancelSndXFTPFile] -> m ()
+    cancelSndXFTPs sndXFTPs = do
+      let xsfIds = map (\(CancelSndXFTPFile FileTransferMeta {fileId} _ xsf) -> (xsf, fileId)) sndXFTPs
+      agentXFTPDeleteSndFilesRemote user xsfIds
+    cancelRcvXFTPs :: [CancelRcvXFTPFile] -> m ()
+    cancelRcvXFTPs rcvXFTPs = do
+      forM_ rcvXFTPs $ \(CancelRcvXFTPFile RcvFileTransfer {fileId} _) ->
+        closeFileHandle fileId rcvFiles `catchChatError` \_ -> pure ()
```
Comment (spaced4ndy) [inline src/Simplex/Chat.hs:2698]: not sure why we only do it for rcv files in cancel

File: src/Simplex/Chat.hs [hunk 2]
```diff
@@ -6225,20 +6297,42 @@ agentXFTPDeleteRcvFile aFileId fileId = do
   withAgent (`xftpDeleteRcvFile` aFileId)
   withStore' $ \db -> setRcvFTAgentDeleted db fileId
 
+agentXFTPDeleteRcvFiles :: ChatMonad m => [(XFTPRcvFile, FileTransferId)] -> m ()
+agentXFTPDeleteRcvFiles rcvFiles = do
+  let rcvFiles' = filter (not . agentRcvFileDeleted . fst) rcvFiles
+  fileIds' <- forM rcvFiles' $ \case
+    (XFTPRcvFile {agentRcvFileId = Just (AgentRcvFileId aFileId)}, fileId) -> do
+      withAgent (`xftpDeleteRcvFile` aFileId)
```
Comment (dpwiz) [inline src/Simplex/Chat.hs:6305]: Good candidate for agent-side batching. Alas, no batching in FileTransfer.Agent yet.

**Review by dpwiz — APPROVED** [review]

File: src/Simplex/Chat.hs [hunk 3]
```diff
@@ -970,37 +972,38 @@ processChatCommand' vr = \case
...
+        deleteUnusedContact db contactId = runExceptT $ withExceptT ChatErrorStore $ do
+          ct <- getContact db user contactId
+          if directOrUsed ct
+            then pure []
+            else
+              liftIO (checkContactHasGroups db user ct) >>= \case
+                Just _ -> pure []
+                Nothing -> do
+                  conns <- liftIO $ getContactConnections db userId ct
+                  setContactDeleted db user ct
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:990]: if this query fails, the connections will not be returned like it was before

File: src/Simplex/Chat.hs [hunk 4]
```diff
@@ -970,37 +972,38 @@ processChatCommand' vr = \case
         withStore' $ \db -> deleteGroupItemsAndMembers db user gInfo members
         withStore' $ \db -> deleteGroup db user gInfo
         let contactIds = mapMaybe memberContactId members
-        deleteAgentConnectionsAsync user . concat =<< mapM deleteUnusedContact contactIds
+        (errs, connIds) <- partitionEithers <$> withStoreBatch (\db -> map (deleteUnusedContact db) contactIds)
+        unless (null errs) $ toView $ CRChatErrors (Just user) errs
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:976]:
```suggestion
        unless (null errs) $ toView $ CRChatErrors (Just user) (map ChatErrorStore errs)
```

File: src/Simplex/Chat.hs [hunk 5]
```diff
@@ -970,37 +972,38 @@ processChatCommand' vr = \case
...
-              | otherwise =
-                  withStore' (\db -> checkContactHasGroups db user ct) >>= \case
-                    Just _ -> pure []
-                    Nothing -> do
-                      conns <- withStore' $ \db -> getContactConnections db userId ct
-                      withStore (\db -> setContactDeleted db user ct)
-                        `catchChatError` (toView . CRChatError (Just user))
-                      pure $ map aConnId conns
+        deleteUnusedContact :: DB.Connection -> ContactId -> IO (Either ChatError [ConnId])
+        deleteUnusedContact db contactId = runExceptT $ withExceptT ChatErrorStore $ do
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:981]:
```suggestion
        deleteUnusedContact :: DB.Connection -> ContactId -> IO (Either StoreError [ConnId])
        deleteUnusedContact db contactId = runExceptT $ do
```

File: src/Simplex/Chat.hs [hunk 6]
```diff
@@ -2556,20 +2562,101 @@ setAllExpireCIFlags b = do
...
-  forM_ filePath $ \fPath ->
-    deleteFileLocally fPath `catchChatError` (toView . CRChatError (Just user))
-  pure aConnIds
+newtype CancelRcvSMPFile = CancelRcvSMPFile RcvFileTransfer
+
+deleteFilesFilesystem :: ChatMonad m => User -> [CIFileInfo] -> m ()
+deleteFilesFilesystem user filesInfo =
+  forM_ filesInfo $ \CIFileInfo {filePath} ->
+    forM_ filePath $ \fPath ->
+      deleteFileLocally fPath `catchChatError` (toView . CRChatError (Just user))
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:2659]: deleteFileLocally never throws error

**Pull request merged**